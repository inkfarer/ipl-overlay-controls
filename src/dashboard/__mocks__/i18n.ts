import i18next from 'i18next';
import { config } from '@vue/test-utils';

jest.mock('i18next-vue', () => ({
    __esModule: true,
    useTranslation: () => ({ t: i18next.t })
}));

beforeAll(async () => {
    await i18next.init({
        lng: 'cimode',
        appendNamespaceToCIMode: true
    });
    config.global.mocks = {
        $t: i18next.t
    };
});
