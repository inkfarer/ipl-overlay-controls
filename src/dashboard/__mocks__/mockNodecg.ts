import Mock = jest.Mock;

export type ReplicantChangeHandler = (newValue?: unknown, oldValue?: unknown) => void;
export type MockReplicant = { value: unknown };
type OnFunction = (event: string, handler: ReplicantChangeHandler) => void;

export class MockNodecg {
    replicants: {[key: string]: MockReplicant};
    listeners: {[key: string]: ReplicantChangeHandler};
    sendMessage: Mock

    constructor() {
        this.replicants = {};
        this.listeners = {};
        this.sendMessage = jest.fn();
    }

    init(): void {
        const self = this;
        window.nodecg = {
            // @ts-ignore: Just a mock
            Replicant(name: string) {
                const replicantValue: { value: unknown, on: OnFunction } = {
                    value: undefined,
                    on: (event: string, handler: ReplicantChangeHandler) => {
                        self.listeners[name] = handler;
                    }
                };
                self.replicants[name] = replicantValue;
                return replicantValue;
            },
            sendMessage: self.sendMessage,
            // @ts-ignore: Just a mock
            log: {
                info: jest.fn()
            }
        };

        // @ts-ignore: It's a mock.
        window.NodeCG = {
            waitForReplicants: jest.fn().mockResolvedValue(true)
        };
    }
}
