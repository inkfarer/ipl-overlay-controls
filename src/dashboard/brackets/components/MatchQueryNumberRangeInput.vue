<template>
    <ipl-radio
        :model-value="props.modelValue ? String(props.modelValue) : null"
        :options="options"
        :label="props.label"
        @update:model-value="newValue => emit('update:modelValue', newValue)"
    />
</template>

<script setup lang="ts">
import { IplRadio } from '@iplsplatoon/vue-components';

const props = defineProps<{
    modelValue?: string | number
    min: number
    max: number
    label: string
}>();
const emit = defineEmits<{
    (event: 'update:modelValue', newValue: string): void
}>();

const options = Array.from(
    { length: props.max - props.min + 1 },
    (_, i) => ({ name: String(i + props.min), value: String(i + props.min) }));
</script>
