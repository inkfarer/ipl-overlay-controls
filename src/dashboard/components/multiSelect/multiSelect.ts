class MultiSelect extends HTMLElement {
    select: HTMLSelectElement;
    selectedElemDisplay: HTMLDivElement;
    selectedOptions: HTMLOptionElement[];

    constructor() {
        super();

        const template = document.createElement('template');
        template.innerHTML = `
            <div class="select-wrapper" id="select-wrapper">
                <slot></slot>
            </div>`;

        this.selectedOptions = [];
        this.select = this.getInnerSelect();
        this.selectedElemDisplay = this.createElemDisplay();

        // Create shadow root with styles and HTML template
        const shadow = this.attachShadow({ mode: 'open' });
        shadow.appendChild(MultiSelect.createStyle());
        shadow.appendChild(template.content.cloneNode(true));
        shadow.getElementById('select-wrapper')
            .appendChild(this.selectedElemDisplay);

        // If select element children change, check if selected elements still exist
        new MutationObserver(() => {
            const newOptions: string[] = Array.from(this.select.options)
                .map((option: HTMLOptionElement) => { return option.value; });
            this.selectedOptions = this.selectedOptions.filter(selectedOption => {
                return newOptions.includes(selectedOption.value);
            });

            this.updateElemDisplay();
            this.select.selectedIndex = -1;
        }).observe(this.select, { childList: true });
    }

    private createElemDisplay(): HTMLDivElement {
        const display = document.createElement('div');
        display.classList.add('elem-display');
        display.addEventListener('click', e => {
            // Delete selected element on click
            if ((e.target as HTMLElement).classList.contains('option')) {
                this.selectedOptions = this.selectedOptions.filter(elem => {
                    const target = e.target as HTMLDivElement;
                    return elem.innerText !== target.innerText &&
                        elem.value !== target.dataset.optionValue;
                });
                this.updateElemDisplay();
                this.dispatchEvent(new Event('change'));
            }
        });
        return display;
    }

    private getInnerSelect(): HTMLSelectElement {
        const select = this.querySelector('select');
        if (!select) {
            throw new Error('No select given to multi-select!');
        }
        select.selectedIndex = -1;

        select.addEventListener('change', event => {
            const selectedOption = select.options[select.selectedIndex];
            if (!this.selectedOptions.includes(selectedOption)) {
                this.selectedOptions.push(selectedOption);
                this.dispatchEvent(new Event('change'));
            }
            select.selectedIndex = -1;
            this.updateElemDisplay();
            event.stopPropagation();
        });

        return select;
    }

    private updateElemDisplay() {
        this.selectedElemDisplay.innerHTML = '';
        this.selectedOptions.forEach(opt => {
            const optionElem = document.createElement('div');
            optionElem.classList.add('option');
            optionElem.innerText = opt.innerText;
            optionElem.dataset.optionValue = opt.value;
            this.selectedElemDisplay.appendChild(optionElem);
        });
    }

    private static createStyle(): HTMLStyleElement {
        const style = document.createElement('style');
        /* eslint-disable max-len */
        style.textContent = `
            ::slotted(select) {
                opacity: 0;
                width: 100%;
                height: 100%;
                position: absolute;
                cursor: pointer;
                z-index: 1;
            }
            
            div.elem-display {
                border-bottom: 1px solid #737373;
                min-height: 26px;
                display: flex;
                flex-wrap: wrap;
                position: relative;
                z-index: 2;
                pointer-events: none;
                padding-bottom: 16px;
                background-image: url("data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'><path fill='none' stroke='white' stroke-linecap='round' stroke-linejoin='round' stroke-width='2' d='M2 5l6 6 6-6'/></svg>");
                background-size: 16px 12px;
                background-repeat: no-repeat;
                background-position: right 3px bottom 5px;
            }
            
            div.elem-display > .option {
                background-color: #2F3A4F;
                border-radius: 8px;
                font-size: 14px;
                padding: 2px 19px 2px 4px;
                margin: 2px;
                pointer-events: auto;
                background-image: url('data:image/svg+xml,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 352 512"><!-- Font Awesome Free 5.15.3 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path fill="%23728EC2" d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"/></svg>');
                background-size: 14px 14px;
                background-repeat: no-repeat;
                background-position: right 3px center;
                transition-duration: 100ms;
            }
            
            div.elem-display > .option:hover {
                background-color: #3C4B66;
            }
            
            div.select-wrapper {
                position: relative;
                cursor: pointer;
            }`;
        /* eslint-enable max-len */

        return style;
    }
}

customElements.define('multi-select', MultiSelect);
