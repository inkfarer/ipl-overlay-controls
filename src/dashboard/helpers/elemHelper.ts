export function appendChildren(element: HTMLElement, ...children: HTMLElement[]): void {
    children.forEach(elem => {
        element.appendChild(elem);
    });
}

export function elementById<T extends HTMLElement>(id: string): T | undefined {
    return document.getElementById(id) as T;
}

export function dispatch(element: EventTarget, event: string): void {
    element.dispatchEvent(new Event(event));
}

export function hideElement(element: HTMLElement): void {
    element.style.display = 'none';
}

export function showElement(element: HTMLElement): void {
    element.style.removeProperty('display');
}

export function prependToBody(elem: HTMLElement): void {
    document.body.prepend(elem);
}

export function appendElementAfter(newElement: HTMLElement, element: HTMLElement): void {
    element.parentNode.insertBefore(newElement, element.nextSibling);
}
