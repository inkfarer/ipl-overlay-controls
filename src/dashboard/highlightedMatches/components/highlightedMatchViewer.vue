<template>
    <ipl-message
        v-if="!matchOptions?.length"
        type="info"
    >
        No matches loaded.
    </ipl-message>
    <ipl-space v-else>
        <ipl-select
            v-model="selectedMatch"
            label="Match"
            :options="matchOptions"
            data-test="match-selector"
        />
        <ipl-data-row
            label="Team A"
            :value="selectedMatchData?.teamA?.name"
            data-test="team-a-name-display"
        />
        <ipl-data-row
            label="Team B"
            :value="selectedMatchData?.teamB?.name"
            data-test="team-b-name-display"
        />
        <ipl-button
            class="m-t-8"
            label="Set next match"
            :disabled="disableSetNextMatch"
            data-test="set-next-match-button"
            @click="handleSetNextMatch"
        />
    </ipl-space>
</template>

<script lang="ts">
import { computed, defineComponent, ref } from 'vue';
import { IplButton, IplSpace, IplSelect, IplDataRow, IplMessage } from '@iplsplatoon/vue-components';
import { useHighlightedMatchStore } from '../highlightedMatchStore';

export default defineComponent({
    name: 'HighlightedMatchViewer',

    components: { IplMessage, IplDataRow, IplSelect, IplSpace, IplButton },

    setup() {
        const store = useHighlightedMatchStore();
        const selectedMatch = ref<string>(null);
        const selectedMatchData = computed(() =>
            store.state.highlightedMatches.find(match => match.meta.id === selectedMatch.value));

        store.watch(store => store.highlightedMatches, newValue => {
            selectedMatch.value = newValue[0]?.meta?.id ?? null;
        }, { immediate: true });

        return {
            matchOptions: computed(() => {
                const isAllSameStage = store.state.highlightedMatches
                    .every(match => match.meta.stageName === store.state.highlightedMatches[0].meta.stageName);

                return store.state.highlightedMatches.map(stage => ({
                    value: stage.meta.id,
                    name: isAllSameStage ? stage.meta.name : `${stage.meta.name} | ${stage.meta.stageName}`
                }));
            }),
            selectedMatch,
            selectedMatchData,
            disableSetNextMatch: computed(() => !selectedMatchData.value),
            async handleSetNextMatch() {
                await store.dispatch('setNextMatch', {
                    teamAId: selectedMatchData.value.teamA.id,
                    teamBId: selectedMatchData.value.teamB.id
                });
            }
        };
    }
});
</script>
