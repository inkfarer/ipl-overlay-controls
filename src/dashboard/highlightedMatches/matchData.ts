import { hideElement, showElement } from '../helpers/elemHelper';
import { addChangeReminder } from '../helpers/buttonHelper';
import { addSelector, clearSelectors } from '../helpers/selectHelper';
import { setImportStatus } from '../helpers/importStatusHelper';
import { ImportStatus } from 'types/enums/importStatus';
import { HighlightedMatches, TournamentData } from 'schemas';
import { Match } from 'types/match';
import { TournamentDataSource } from 'types/enums/tournamentDataSource';
import { BracketType } from 'types/enums/bracketType';
import { SetRoundRequest } from 'types/messages/rounds';
import { MultiSelect } from '../components/multiSelect';
import { addDots } from '../helpers/stringHelper';

const highlightedMatchData = nodecg.Replicant<HighlightedMatches>('highlightedMatches');
const tournamentData = nodecg.Replicant<TournamentData>('tournamentData');

const matchDataStatusElem = document.getElementById('match-data-status');
const stageSelectElem = document.getElementById('stage-selector') as MultiSelect;
const matchSelectElem = document.getElementById('match-selector') as HTMLSelectElement;
const setNextMatchBtn = document.getElementById('set-next-match-btn') as HTMLButtonElement;
const getMatchesBtn = document.getElementById('get-matches') as HTMLButtonElement;

const unsupportedPlatformWarning = document.getElementById('unsupported-service-message');
const noLoadedMatchesMessage = document.getElementById('load-matches-hint');
const loadMatchesSpace = document.getElementById('load-matches-space');
const selectMatchSpace = document.getElementById('select-match-space');

const teamAName = document.getElementById('team-a-name');
const teamBName = document.getElementById('team-b-name');

/**
 * Gets match from matchID
 * @param id the ID of the match
 */
function getMatchFromID(id: string): Match | null {
    return highlightedMatchData.value.filter(match => match.meta.id === id)[0];
}

// When get match button is pressed
getMatchesBtn.addEventListener('click', () => {
    setImportStatus(ImportStatus.LOADING, matchDataStatusElem);
    const selectedStages: string[] = stageSelectElem.selectedOptions.map(option => { return option.value; });

    if (selectedStages.length > 0) {
        // Send message to extension
        nodecg.sendMessage('getHighlightedMatches', {
            stages: selectedStages,
            getAllStages: selectedStages.includes('AllStages')
        }, (e, result) => {
            // If we get an error
            if (e) {
                console.error(e);
                setImportStatus(ImportStatus.FAILURE, matchDataStatusElem);
                return;
            }
            // If we get success
            setImportStatus(result.status, matchDataStatusElem);
        });
    }
});

stageSelectElem.addEventListener('change', event => {
    getMatchesBtn.disabled = (event.target as MultiSelect).selectedOptions.length < 1;
});

// When set next match button is pressed
setNextMatchBtn.addEventListener('click', () => {
    const selectedMatch = getMatchFromID(matchSelectElem.value);
    if (selectedMatch) {
        nodecg.sendMessage('setNextRound',
            { teamAId: selectedMatch.teamA.id, teamBId: selectedMatch.teamB.id } as SetRoundRequest);
    }
});

tournamentData.on('change', newValue => {
    clearSelectors('stage-selector');
    if (canImportDataFrom(newValue.meta.source as TournamentDataSource)) {
        showElement(loadMatchesSpace);
        hideElement(unsupportedPlatformWarning);

        for (let i = 0; i < newValue.stages.length; i++) {
            const element = newValue.stages[i];
            if ([BracketType.SWISS,
                BracketType.DOUBLE_ELIMINATION,
                BracketType.SINGLE_ELIMINATION,
                BracketType.ROUND_ROBIN].includes(element.type as BracketType)) {

                // if bracket type is supported, add to drop down
                addSelector(addDots(element.name), 'stage-selector', element.id);
            }
        }
        addSelector('All Brackets', 'stage-selector', 'AllStages');
    } else {
        hideElement(loadMatchesSpace);
        showElement(unsupportedPlatformWarning);
        hideElement(noLoadedMatchesMessage);
    }
});

NodeCG.waitForReplicants(tournamentData).then(() => {
    highlightedMatchData.on('change', newValue => {
        clearSelectors('match-selector');
        if (!newValue || newValue.length < 1) {
            hideElement(selectMatchSpace);
            if (canImportDataFrom(tournamentData.value.meta.source as TournamentDataSource)) {
                showElement(noLoadedMatchesMessage);
            }
        } else {
            hideElement(noLoadedMatchesMessage);
            showElement(selectMatchSpace);

            const isSameBracket = newValue.every(value => {
                return value.meta.stageName === newValue[0].meta.stageName;
            });

            // fill drop down with matches
            newValue.forEach(function (value) {
                const matchName = isSameBracket ? value.meta.name : `${value.meta.name} | ${value.meta.stageName}`;
                addSelector(
                    addDots(matchName),
                    'match-selector',
                    value.meta.id);
            });
            teamAName.innerText = addDots(newValue[0].teamA.name);
            teamBName.innerText = addDots(newValue[0].teamB.name);
        }
    });
});

matchSelectElem.addEventListener('input', () => {
    const selectedMatch = getMatchFromID(matchSelectElem.value);
    if (selectedMatch) {
        teamAName.innerText = selectedMatch.teamA.name;
        teamBName.innerText = selectedMatch.teamB.name;
        setNextMatchBtn.disabled = false;
    } else {
        teamAName.innerText = 'Unknown';
        teamBName.innerText = 'Unknown';
        setNextMatchBtn.disabled = true;
    }
});

addChangeReminder(document.querySelectorAll('.next-match-update-warning'), setNextMatchBtn);

function canImportDataFrom(source: TournamentDataSource): boolean {
    return [TournamentDataSource.BATTLEFY].includes(source);
}
