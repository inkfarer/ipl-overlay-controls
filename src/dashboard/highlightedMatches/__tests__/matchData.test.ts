import { MockNodecg } from '../../__mocks__/mockNodecg';
import { dispatch, elementById } from '../../helpers/elemHelper';
import { MultiSelect } from '../../components/multiSelect';
import { TournamentDataSource } from 'types/enums/tournamentDataSource';
import { BracketType } from 'types/enums/bracketType';
import difference from 'lodash/difference';

describe('matchData', () => {
    let nodecg: MockNodecg;

    beforeEach(() => {
        jest.resetModules();
        jest.resetAllMocks();

        nodecg = new MockNodecg();
        nodecg.init();

        document.body.innerHTML = `
            <div id="match-data-status"></div>
            <multi-select id="stage-selector">
                <select class="stage-selector" id="inner-stage-selector"></select>
            </multi-select>
            <select id="match-selector" class="match-selector"></select>
            <button id="set-next-match-btn"></button>
            <button id="get-matches"></button>
            <label id="match-label"></label>
            <div id="unsupported-service-message"></div>
            <div id="load-matches-hint"></div>
            <div id="load-matches-space"></div>
            <div id="select-match-space"></div>
            <div id="team-a-name"></div>
            <div id="team-b-name"></div>`;

        require('../matchData');
    });

    describe('get-matches: click', () => {
        describe('source: BATTLEFY', () => {
            it('sends a message if getting specific stages', () => {
                nodecg.replicants.tournamentData.value = { meta: { source: TournamentDataSource.BATTLEFY } };
                elementById<MultiSelect>('stage-selector').selectedOptions
                    = [{ value: 'stage-1' }, { value: 'stage-2' }] as unknown as HTMLOptionElement[];

                dispatch(elementById('get-matches'), 'click');

                expect(nodecg.sendMessage).toHaveBeenCalledWith('getHighlightedMatches', {
                    stages: [ 'stage-1', 'stage-2' ],
                    getAllMatches: false
                },
                expect.any(Function));
            });

            it('sends a message if getting all stages', () => {
                nodecg.replicants.tournamentData.value = { meta: { source: TournamentDataSource.BATTLEFY } };
                elementById<MultiSelect>('stage-selector').selectedOptions
                    = [{ value: 'stage-1' }, { value: 'AllStages' }] as unknown as HTMLOptionElement[];

                dispatch(elementById('get-matches'), 'click');

                expect(nodecg.sendMessage).toHaveBeenCalledWith('getHighlightedMatches', {
                    stages: [ 'stage-1', 'AllStages' ],
                    getAllMatches: true
                },
                expect.any(Function));
            });

            it('sends no message if no stages have been selected', () => {
                nodecg.replicants.tournamentData.value = { meta: { source: TournamentDataSource.BATTLEFY } };
                elementById<MultiSelect>('stage-selector').selectedOptions = [] as unknown as HTMLOptionElement[];

                dispatch(elementById('get-matches'), 'click');

                expect(nodecg.sendMessage).not.toHaveBeenCalled();
            });
        });

        describe('source: SMASHGG', () => {
            it('sends a message if getting specific stages', () => {
                nodecg.replicants.tournamentData.value = { meta: { source: TournamentDataSource.SMASHGG } };
                elementById<MultiSelect>('stage-selector').selectedOptions
                    = [{ value: '123123' }, { value: '345345' }] as unknown as HTMLOptionElement[];

                dispatch(elementById('get-matches'), 'click');

                expect(nodecg.sendMessage).toHaveBeenCalledWith(
                    'getHighlightedMatches',
                    { streamIDs: [ 123123, 345345 ], getAllMatches: false },
                    expect.any(Function));
            });

            it('sends a message if getting all stages', () => {
                nodecg.replicants.tournamentData.value = { meta: { source: TournamentDataSource.SMASHGG } };
                elementById<MultiSelect>('stage-selector').selectedOptions
                    = [{ value: '234' }, { value: 'AllStreams' }] as unknown as HTMLOptionElement[];

                dispatch(elementById('get-matches'), 'click');

                expect(nodecg.sendMessage).toHaveBeenCalledWith('getHighlightedMatches', {
                    streamIDs: [],
                    getAllMatches: true
                },
                expect.any(Function));
            });

            it('sends no message if no stages have been selected', () => {
                nodecg.replicants.tournamentData.value = { meta: { source: TournamentDataSource.SMASHGG } };
                elementById<MultiSelect>('stage-selector').selectedOptions = [] as unknown as HTMLOptionElement[];

                dispatch(elementById('get-matches'), 'click');

                expect(nodecg.sendMessage).not.toHaveBeenCalled();
            });
        });
    });

    describe('stage-selector: change', () => {
        it('disables get matches button if no options are selected', () => {
            const selector = elementById<MultiSelect>('stage-selector');
            selector.selectedOptions = [] as unknown as HTMLOptionElement[];

            dispatch(selector, 'change');

            expect(elementById<HTMLButtonElement>('get-matches').disabled).toEqual(true);
        });

        it('enables get matches button if options are selected', () => {
            const selector = elementById<MultiSelect>('stage-selector');
            selector.selectedOptions = [{}] as unknown as HTMLOptionElement[];

            dispatch(selector, 'change');

            expect(elementById<HTMLButtonElement>('get-matches').disabled).toEqual(false);
        });
    });

    describe('set-next-match-btn: click', () => {
        it('sends a message to set round data', () => {
            const matchSelector = elementById<HTMLSelectElement>('match-selector');
            matchSelector.innerHTML = `
                <option value="round1">r1</option>
                <option value="round2">r2</option>`;
            matchSelector.value = 'round2';
            nodecg.replicants.highlightedMatches.value = [
                {
                    meta: { id: 'round1' }
                },
                {
                    meta: { id: 'round2' },
                    teamA: { id: 'aaaa' },
                    teamB: { id: 'bbbb' }
                }
            ];

            dispatch(elementById('set-next-match-btn'), 'click');

            expect(nodecg.sendMessage).toHaveBeenCalledWith('setNextRound', {
                teamAId: 'aaaa',
                teamBId: 'bbbb'
            });
        });

        it('sends no message if the round is not found', () => {
            const matchSelector = elementById<HTMLSelectElement>('match-selector');
            matchSelector.innerHTML = `
                <option value="round1">r1</option>
                <option value="round2">r2</option>`;
            matchSelector.value = 'round2';
            nodecg.replicants.highlightedMatches.value = [
                {
                    meta: { id: 'round1' }
                },
                {
                    meta: { id: 'round3' },
                    teamA: { id: 'aaaa' },
                    teamB: { id: 'bbbb' }
                }
            ];

            dispatch(elementById('set-next-match-btn'), 'click');

            expect(nodecg.sendMessage).not.toHaveBeenCalled();
        });
    });

    describe('tournamentData: change', () => {
        it('hides messages and updates stages if source is BATTLEFY', () => {
            const stageSelector = elementById<HTMLSelectElement>('inner-stage-selector');
            stageSelector.innerHTML = '<option value="old-option"></option>';

            nodecg.listeners.tournamentData({
                meta: { source: TournamentDataSource.BATTLEFY },
                stages: [
                    { type: BracketType.DOUBLE_ELIMINATION, name: 'Cool Stage', id: '123123' },
                    {
                        type: BracketType.LADDER,
                        name: 'Ladder Stage Ladder Stage Ladder Stage Ladder Stage Ladder Stage Ladder Stage',
                        id: '345345'
                    },
                    { type: BracketType.SINGLE_ELIMINATION, name: 'Cool Stage 2', id: '5467567' },
                    { type: BracketType.ROUND_ROBIN, name: 'Robin Stage', id: '1236' },
                    { type: BracketType.SWISS, name: 'Swiss Stage', id: '9999' }
                ]
            });

            expect(stageSelector.innerHTML).toMatchSnapshot();
            expect(elementById('load-matches-space').style.display).toEqual('');
            expect(elementById('unsupported-service-message').style.display).toEqual('none');
            expect(elementById('match-label').innerText).toEqual('Bracket');
        });

        it('hides messages and updates stages if source is SMASHGG', () => {
            const stageSelector = elementById<HTMLSelectElement>('inner-stage-selector');
            stageSelector.innerHTML = '<option value="old-option"></option>';

            nodecg.listeners.tournamentData({
                meta: {
                    source: TournamentDataSource.SMASHGG,
                    sourceSpecificData: {
                        smashgg: {
                            streams: [
                                { streamName: 'Cool Stream', id: 123123 },
                                { streamName: 'Cool Stream 2', id: 345345 }
                            ]
                        }
                    }
                }
            });

            expect(stageSelector.innerHTML).toMatchSnapshot();
            expect(elementById('load-matches-space').style.display).toEqual('');
            expect(elementById('unsupported-service-message').style.display).toEqual('none');
            expect(elementById('match-label').innerText).toEqual('Stream');
        });

        difference(
            Object.values(TournamentDataSource),
            [TournamentDataSource.BATTLEFY, TournamentDataSource.SMASHGG]
        ).forEach(source => {
            it(`shows warning message if source is ${source}`, () => {
                const stageSelector = elementById<HTMLSelectElement>('inner-stage-selector');
                stageSelector.innerHTML = '<option value="old-option"></option>';

                nodecg.listeners.tournamentData({
                    meta: { source: source },
                });

                expect(stageSelector.innerHTML).toEqual('');
                expect(elementById('load-matches-space').style.display).toEqual('none');
                expect(elementById('unsupported-service-message').style.display).toEqual('');
                expect(elementById('load-matches-hint').style.display).toEqual('none');
            });
        });
    });

    describe('highlightedMatchData: change', () => {
        describe('no matches available', () => {
            it('displays message if source is BATTLEFY', () => {
                const loadMatchesMsg = elementById('load-matches-hint');
                loadMatchesMsg.style.display = 'none';
                nodecg.replicants.tournamentData.value = { meta: { source: TournamentDataSource.BATTLEFY } };

                nodecg.listeners.highlightedMatches([]);

                expect(elementById('select-match-space').style.display).toEqual('none');
                expect(loadMatchesMsg.style.display).toEqual('');
            });

            it('displays message if source is SMASHGG', () => {
                const loadMatchesMsg = elementById('load-matches-hint');
                loadMatchesMsg.style.display = 'none';
                nodecg.replicants.tournamentData.value = { meta: { source: TournamentDataSource.SMASHGG } };

                nodecg.listeners.highlightedMatches([]);

                expect(elementById('select-match-space').style.display).toEqual('none');
                expect(loadMatchesMsg.style.display).toEqual('');
            });

            difference(
                Object.values(TournamentDataSource),
                [TournamentDataSource.BATTLEFY, TournamentDataSource.SMASHGG]
            ).forEach(source => {
                it(`displays no match selector or message if source is ${source}`, () => {
                    const loadMatchesMsg = elementById('load-matches-hint');
                    loadMatchesMsg.style.display = 'none';
                    nodecg.replicants.tournamentData.value = { meta: { source: source } };

                    nodecg.listeners.highlightedMatches([]);

                    expect(elementById('select-match-space').style.display).toEqual('none');
                    expect(loadMatchesMsg.style.display).toEqual('none');
                });
            });
        });

        describe('with matches', () => {
            it('hides message', () => {
                const selectMatchSpace = elementById('select-match-space');
                selectMatchSpace.style.display = 'none';

                nodecg.listeners.highlightedMatches([{
                    meta: { stageName: 'Cool Stage' },
                    teamA: { name: 'Cool Team' },
                    teamB: { name: 'Cooler Team' }
                }]);

                expect(elementById('load-matches-hint').style.display).toEqual('none');
                expect(selectMatchSpace.style.display).toEqual('');
            });

            it('fills match selector if all matches are on the same stage', () => {
                nodecg.listeners.highlightedMatches([
                    {
                        meta: { stageName: 'Cool Stage', id: '111', name: 'Match 1' },
                        teamA: { name: 'Cool Team' },
                        teamB: { name: 'Cooler Team' }
                    },
                    {
                        meta: { stageName: 'Cool Stage', id: '222', name: 'Match 2' },
                        teamA: { name: 'Cool Team 2' },
                        teamB: { name: 'Cooler Team 2' }
                    },
                    {
                        meta: { stageName: 'Cool Stage', id: '333', name: 'Match 3' },
                        teamA: { name: 'Cool Team 3' },
                        teamB: { name: 'Cooler Team 3' }
                    }
                ]);

                expect(elementById('match-selector').innerHTML).toMatchSnapshot();
                expect(elementById('team-a-name').innerText).toEqual('Cool Team');
                expect(elementById('team-b-name').innerText).toEqual('Cooler Team');
            });

            it('fills match selector if all matches are not on the same stage', () => {
                nodecg.listeners.highlightedMatches([
                    {
                        meta: { stageName: 'Cool Stage', id: '123', name: 'Dope Match 1' },
                        teamA: { name: 'Dope Team' },
                        teamB: { name: 'Doper Team' }
                    },
                    {
                        meta: { stageName: 'Cool Stage 2', id: '234', name: 'Cool Match 2' },
                        teamA: { name: 'Cool Team' },
                        teamB: { name: 'Cooler Team' }
                    },
                    {
                        meta: { stageName: 'Cool Stage 3', id: '345', name: 'Rad Match 3' },
                        teamA: { name: 'Rad Team' },
                        teamB: { name: 'Radder team' }
                    }
                ]);

                expect(elementById('match-selector').innerHTML).toMatchSnapshot();
                expect(elementById('team-a-name').innerText).toEqual('Dope Team');
                expect(elementById('team-b-name').innerText).toEqual('Doper Team');
            });
        });
    });

    describe('match-selector: input', () => {
        it('sets match data if found', () => {
            const selector = elementById<HTMLSelectElement>('match-selector');
            selector.innerHTML = `
                <option value="123123">
                <option value="234234">`;
            selector.value = '234234';
            nodecg.replicants.highlightedMatches.value = [
                { meta: { id: '123123' }, teamA: { name: 'Cool Team' }, teamB: { name: 'Cool Team 2' } },
                { meta: { id: '234234' }, teamA: { name: 'Rad Team' }, teamB: { name: 'Rad Team 2' } }
            ];

            dispatch(selector, 'input');

            expect(elementById('team-a-name').innerText).toEqual('Rad Team');
            expect(elementById('team-b-name').innerText).toEqual('Rad Team 2');
            expect(elementById<HTMLButtonElement>('set-next-match-btn').disabled).toEqual(false);
        });

        it('handles missing match data', () => {
            const selector = elementById<HTMLSelectElement>('match-selector');
            selector.innerHTML = `
                <option value="123123">
                <option value="456456">`;
            selector.value = '456456';
            nodecg.replicants.highlightedMatches.value = [
                { meta: { id: '123123' }, teamA: { name: 'Cool Team' }, teamB: { name: 'Cool Team 2' } },
                { meta: { id: '234234' }, teamA: { name: 'Rad Team' }, teamB: { name: 'Rad Team 2' } }
            ];

            dispatch(selector, 'input');

            expect(elementById('team-a-name').innerText).toEqual('Unknown');
            expect(elementById('team-b-name').innerText).toEqual('Unknown');
            expect(elementById<HTMLButtonElement>('set-next-match-btn').disabled).toEqual(true);
        });
    });
});
