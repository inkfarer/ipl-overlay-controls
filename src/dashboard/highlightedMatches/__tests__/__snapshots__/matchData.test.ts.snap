// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchData highlightedMatchData: change with matches fills match selector if all matches are not on the same stage 1`] = `"<option value=\\"123\\">Dope Match 1 | Cool Stage</option><option value=\\"234\\">Cool Match 2 | Cool Stage 2</option><option value=\\"345\\">Rad Match 3 | Cool Stage 3</option>"`;

exports[`matchData highlightedMatchData: change with matches fills match selector if all matches are on the same stage 1`] = `"<option value=\\"111\\">Match 1</option><option value=\\"222\\">Match 2</option><option value=\\"333\\">Match 3</option>"`;

exports[`matchData tournamentData: change hides messages and updates stages if source is BATTLEFY 1`] = `"<option value=\\"123123\\">Cool Stage</option><option value=\\"5467567\\">Cool Stage 2</option><option value=\\"1236\\">Robin Stage</option><option value=\\"9999\\">Swiss Stage</option><option value=\\"AllStages\\">All Brackets</option>"`;

exports[`matchData tournamentData: change hides messages and updates stages if source is SMASHGG 1`] = `"<option value=\\"123123\\">Cool Stream</option><option value=\\"345345\\">Cool Stream 2</option><option value=\\"AllStreams\\">All Streams</option>"`;
